File: leet_code_solutions.js/15.js
// Solved.
// Runtime: better than 73.9% of JS solutions
// Memory Usage: better than 98.99 % of JS solutions

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  nums.sort((a, b) => a - b);
  var returnArray = [];
  var start;
  var end;
  var targetSum;

  for (let i = 0; i < nums.length; i++) {
    if (nums[i] === nums[i - 1]) {
      continue;
    }
    targetSum = -nums[i];
    start = i + 1;
    end = nums.length - 1;

    while (start < end) {
      let sum = nums[start] + nums[end];
      if (sum === targetSum) {
        returnArray.push([-targetSum, nums[start], nums[end]]);
        start++;
        end--;
        while (start < end && nums[start] === nums[start - 1]) {
          start++;
        }
        while (start < end && nums[end] === nums[end + 1]) {
          end--;
        }
      } else if (sum > targetSum) {
        end--;
      } else if (sum < targetSum) {
        start++;
      }
    }
  }
  return returnArray;
};

console.log(threeSum([-1, 0, 1, 2, -1, -4]));

File: leet_code_solutions.py/15.py

def threeSum( nums: List[int]) -> List[List[int]]:
    nums.sort()
    ans=[]

    for i in range(len(nums)-2):
            if (i==0) or not (nums[i]==nums[i-1]):
                    low=i+1
                    high=len(nums)-1
                    k=-(nums[i])
                    while(high>low):
                            if (nums[low]+nums[high])==k:
                                    ans.append([nums[i],nums[low],nums[high]])
                                    while(low<high and nums[low]==nums[low+1]): low+=1
                                    while(low<high and nums[high]==nums[high-1]):high-=1
                                    high-=1
                                    low+=1
                            elif (nums[low]+nums[high]>k):high-=1
                            else: low+=1
    return ans
print(threeSum([[-1,0,1,2,-1,-4]))
